common/text/syntax_tree_context.h

```diff
@@ -31,38 +33,16 @@ namespace verible {
 // Note: Public methods are named to follow STL convention for std::stack.
 // TODO(fangism): implement a general ForwardMatcher and ReverseMatcher
 // interface that can express AND/OR/NOT.
-class SyntaxTreeContext {
+class SyntaxTreeContext : public AutoPopStack<const SyntaxTreeNode*> {
  public:
   // member class to handle push and pop of stack safely
-  class AutoPop {
-   public:
-    AutoPop(SyntaxTreeContext*, const SyntaxTreeNode& node);  // pushes
-    ~AutoPop();                                               // pops
-   private:
-    SyntaxTreeContext* context_;
-  };
-
-  // All elements of stack are non-null because they refer to traversed nodes.
-  typedef std::vector<const SyntaxTreeNode*> stack_type;
-  typedef stack_type::const_iterator const_iterator;
-  typedef stack_type::const_reverse_iterator const_reverse_iterator;
-
-  // returns depth of context stack
-  size_t size() const { return stack_.size(); }
-
-  // returns true if the stack is empty
-  bool empty() const { return stack_.empty(); }
+  typedef AutoPopStack<const SyntaxTreeNode*> base_type;
+  using AutoPop = base_type::AutoPop;
 
   // returns the top SyntaxTreeNode of the stack
-  const SyntaxTreeNode& top() const;
-
-  // Allow read-only random-access into stack:
-  const_iterator begin() const { return stack_.begin(); }
-  const_iterator end() const { return stack_.end(); }
-
-  // Reverse iterators be useful for searching from the top-of-stack downward.
-  const_reverse_iterator rbegin() const { return stack_.rbegin(); }
-  const_reverse_iterator rend() const { return stack_.rend(); }
+  const SyntaxTreeNode& top() const {
+    return *ABSL_DIE_IF_NULL(base_type::top());
```

Comment:
You can hide the base class's interface with:

```c++
protected:
  using base_type::top;
```

This makes it clear that you are further restricting direct uses of the base class methods intentionally, and here is where you can alter the interface to return references if you wish.  

Optionally, you can also be explicit about what base class interfaces you wish to expose:

```c++
public:
  using base_type::begin;
```

Reply:
Just restricted access to `base_type::top`. Also removed `base_type::` from methods like `rbegin`, `rend` and so on.